### ISSUE TITLE
[Bug]: Data nodes crashing with error: empty grpc client: context canceled in milvus version 2.3.9

### ISSUE THREAD

**Issue by kdabbir** [milvus-io/milvus#30905]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Environment

```markdown
- Milvus version: 2.3.9
- Deployment mode(standalone or cluster): Milvus cluster deployed in kubernetes
- MQ type(rocksmq, pulsar or kafka):  External Kafka (AWS MSK)
- SDK version(e.g. pymilvus v2.0.0rc2): pymilvus 2.3.6	
num_shards of collection: 1 
total of 1000 collections with 1 shards
Vector dim: 768
Scale of upto 1 billion
- CPU/Memory:
Worker nodes: Data nodes - 15 CPU/54GB memory, Index nodes - 15 CPU/25GB Mem, Query nodes - 14 CPU/118GB
Referred to milvus sizing tool & scaled up 8 index nodes, 8 data nodes, 6 proxy & 25 query nodes.
coordinator nodes:  8 core cpu/16gb memory, Based on c5a.2xlarge.

- Also increased etcd ETCD_MAX_TXN_OPS to 10000 to support more operations
```

### Current Behavior

Follow-up to the [previous issue ](https://github.com/milvus-io/milvus/issues/30004) we've raised, we re-ran our tests with 2 milvus clusters having setting:
1) Normal Insert 2) Bulk Insert

 Have created 1000 collections (no partitions created) with 1 shard and started ingesting 500k records per collection.
 Every insert call has 100k records and we insert randomly into these 1000 collections till they reach 500k records.

After we inserted around 20 Million vectors we saw errors in data nodes for both bulk & normal insert throwing exceptions `rpc error: code = DeadlineExceeded desc = context deadline exceeded` , `retry func failed rpc error: code = DeadlineExceeded desc = context deadline exceeded` and `failed to update channel checkpoint`

I've attached the current log export for both normal insert and bulk insert. Please help understand the issue and suggest a fix to this.

### Expected Behavior

Cluster should be able to work with 1 billion vectors for 1000 collections

### Milvus Log

[Bulk Insert Milvus Error Export Logs.zip](https://github.com/milvus-io/milvus/files/14477191/Bulk.Insert.Milvus.Error.Export.Logs.zip)
[Standard Insert Milvus Error Export Logs.zip](https://github.com/milvus-io/milvus/files/14477192/Standard.Insert.Milvus.Error.Export.Logs.zip)

**Comment by yanliang567** [comment]
/assign @MrPresent-Han 
could u please help to take a look?
/unassign

**Comment by kdabbir** [comment]
@yanliang567 @xiaofan-luan would appreciate your inputs on this issue. We are unable to run even 1000 collections with 2.3.9. I've attached log export of current state of cluster. If you need more logs I can gather them from splunk and provide them when the errors started occurring

**Comment by MrPresent-Han** [comment]
I will handle this

**Comment by kdabbir** [comment]
@MrPresent-Han  yes we have done with export-log, we have some extra metric blocks which may come in the log, you need to scroll down in the log and rest of the log will have the actual milvus log.

**Comment by kdabbir** [comment]
@MrPresent-Han let me re-export the logs not sure why it exported only metrics

**Comment by kdabbir** [comment]
@MrPresent-Han I've reuploaded the logs in the original post/comment. I had to export logs from splunk as there was an issue with k9s export. But I made sure all respective pod logs are expored

**Comment by yiwangdr** [comment]
@kdabbir 
I took a quick look. It looks like the coordinators were overwhelmed by certain requests/tasks.
Could you check the the resource usages (CPU & Memory) of coordinators during those time periods?

**Comment by kdabbir** [comment]
@yiwangdr I see data coord taking upto 12GB memory in spikes during this period. Why does data coord need so much memory? Other nodes metrics seems fine in terms of cpu/memory. Attaching snapshots of our graphana.

Data coord:
<img width="1642" alt="Screenshot 2024-03-04 at 4 41 20 PM" src="https://github.com/milvus-io/milvus/assets/29181508/07576f71-4307-47c3-93bf-79277aa4b1fd">

Data worker:
<img width="1658" alt="Screenshot 2024-03-04 at 4 41 26 PM" src="https://github.com/milvus-io/milvus/assets/29181508/806452ef-7c5c-4b30-a727-6f063f0858b5">

Index-coord:
<img width="1673" alt="Screenshot 2024-03-04 at 4 41 32 PM" src="https://github.com/milvus-io/milvus/assets/29181508/a985dcf1-70bc-4fad-9528-ad2be1da8877">

Index-worker:
<img width="1659" alt="Screenshot 2024-03-04 at 4 41 38 PM" src="https://github.com/milvus-io/milvus/assets/29181508/56b8a2dd-b64a-4989-8b2a-0d2a920eb811">

proxy-worker:
<img width="1650" alt="Screenshot 2024-03-04 at 4 41 46 PM" src="https://github.com/milvus-io/milvus/assets/29181508/38100b42-c7c4-4e48-80f3-d8720cd30fff">

Query-coord:
<img width="1662" alt="Screenshot 2024-03-04 at 4 41 54 PM" src="https://github.com/milvus-io/milvus/assets/29181508/dd1ae365-863b-417a-8ec1-1b6c0c02c375">

Query-worker:
<img width="1691" alt="Screenshot 2024-03-04 at 4 41 59 PM" src="https://github.com/milvus-io/milvus/assets/29181508/c89e097f-f3d8-4098-9022-bd03bed6f3fa">

root-coord:
<img width="1659" alt="Screenshot 2024-03-04 at 4 42 22 PM" src="https://github.com/milvus-io/milvus/assets/29181508/d1420d9d-2ea5-438e-9c39-09e03a948d52">

**Comment by yiwangdr** [comment]
@kdabbir Yes, the datacoord memory usage pattern looks bizarre. We can't reproduce it.
Could you pprof it? Milvus has pprof enabled by default at port 9091.

sample commands: 
curl http://{ip_of_datacoord}:9091/debug/pprof/heap > heap.out
curl http://{ip_of_datacoord}:9091/debug/pprof/goroutine > goroutine.out

**Comment by xiaofan-luan** [comment]
most likely it is due to some bottleneck of datacoord meta. 
See #30837, #30823
Since there is a single lock on meta operation, many of the datacoord request might be blocked by the lock

**Comment by xiaofan-luan** [comment]
a pprof could really helpful.
From the current info, datacoord definitely went timeout
![image](https://github.com/milvus-io/milvus/assets/83447078/631ee4d6-dbde-4fbc-9fda-5a6ca2dfb10d)

**Comment by yanliang567** [comment]
@kdabbir I am trying to reproduce this issue in house, but it seems that I met a different situation with you. I did not see much memory increase on datacoord pod, while I see much cpu and memory increase on datanode pods. So if i convenient, could you please share some code snippets for reproduce this issue?
/assign @kdabbir

**Comment by kdabbir** [comment]
@yanliang567 @MrPresent-Han  can we have a call to discuss on this? I can show my test setup and show the logs/dashboard of the issue. Let me know what time works tomorrow for you. Test code might not give a full picture on this issue.

**Comment by kdabbir** [comment]
@jaime0815 @xiaofan-luan I had debugged into this flow and I see UpdateChannelCheckPoint meta lock from datacoord is causing issues in this case, in the bug I've reported. I have seen this PR is closed
- Remove the global lock from channel checkpoint operations https://github.com/milvus-io/milvus/pull/30903
Can we know why this is closed? This could fix the issue we are seeing here

**Comment by kdabbir** [comment]
| Could you pprof it when the memory usage is high? Milvus has pprof enabled by default at port 9091.

@xiaofan-luan @yanliang567 @MrPresent-Han 
Attaching pprof of goroutune and heap when memory usage for datacoord is around 8GB.
[goroutine and heap compressed.zip](https://github.com/milvus-io/milvus/files/14521272/goroutine.and.heap.compressed.zip)

Attached python milvus SDK scale script we are using:
[milvus_scale_script.py.zip](https://github.com/milvus-io/milvus/files/14521319/milvus_scale_script.py.zip)

**Comment by MrPresent-Han** [comment]
I will take a look at this

**Comment by MrPresent-Han** [comment]
@kdabbir, this problem is still because of global lock, another pr:
https://github.com/milvus-io/milvus/pull/30941 will fix this error

**Comment by kdabbir** [comment]
thanks. when can we expect a release for this? would like to test this fix out and report back

**Comment by MrPresent-Han** [comment]
> thanks. when can we expect a release for this? would like to test this fix out and report back

the version 2.3.11 is on the way,
next Tuesday or Wednesday, I suppose,

**Comment by yanliang567** [comment]
Milvus v2.3.11 was released last nightly, please enjoy it.:P

**Comment by xiaofan-luan** [comment]
/assign @yanliang567 
mark it as fixed?
